### PR TITLE
Add initial implementation and conversion of functional case tasks

### PR DESCRIPTION
**Pull request by mapingo**
### Change description ###
Add implementation of functional case tasks, that are reduced and applied to a CaseDetails instance.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/divorce/caseworker/event/CaseworkerIssueApplication.java [hunk 1]
```diff
@@ -67,13 +62,13 @@ public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuild
...
-                    .mandatory(MarriageDetails::getPlaceOfMarriage)
-                    .done()
-                .done();
+            .complex(Application::getMarriageDetails)
+            .optional(MarriageDetails::getDate)
+            .optional(MarriageDetails::getApplicant1Name)
+            .optional(MarriageDetails::getApplicant2Name)
+            .mandatory(MarriageDetails::getPlaceOfMarriage)
+            .done()
+            .done();
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/divorce/caseworker/event/CaseworkerIssueApplication.java:71]: formatting ?

Comment (mapingo) [reply src/main/java/uk/gov/hmcts/divorce/caseworker/event/CaseworkerIssueApplication.java]: Intellij keeps on doing that, very annoying

File: src/main/java/uk/gov/hmcts/divorce/divorcecase/task/CaseTask.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.function.Function;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Stream.of;
+
+public interface CaseTask extends Function<CaseDetails<CaseData, State>, CaseDetails<CaseData, State>> {
+
+    @SafeVarargs
+    static CaseTaskRunner caseTasks(final Function<CaseDetails<CaseData, State>, CaseDetails<CaseData, State>>... tasks) {
```
Comment (mapingo) [inline src/main/java/uk/gov/hmcts/divorce/divorcecase/task/CaseTask.java:15]: This could be inside CaseTaskRunner

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/divorce/divorcecase/task/CaseTask.java:15]: yeah would be nice I think if we can move it there

**Review by linusnorton — APPROVED** [review]

**Pull request merged**